feat: acknowledge OOO request

surajmaity1 opened the pull request:
<!-- Date Here in dd mmm yyyy-->
Date: Date: 23 Feb, 2025
<!--Developer Name Here-->
Developer Name: Suraj Maity ( surajmaity981@gmail.com )

---

## Issue Ticket Number
<!--Issue ticket this PR closes-->
- Open #2242 
## Description
- [User OOO Management Design Document](https://docs.google.com/document/d/1c9crE6DqLBn46aO9kU3PF5ctTPdhF1CpeatmpgazmIQ/edit?usp=sharing)
<!--Description of the changes made in this PR-->

### Documentation Updated?

- [ ] Yes
- [x] No

<!--Additional notes about documentation update if applicable-->

### Under Feature Flag

- [ ] Yes
- [ ] No

<!--Indicate if changes are under a feature flag-->

### Database Changes

- [ ] Yes
- [ ] No

<!--Notes on any database changes-->

### Breaking Changes

- [ ] Yes
- [x] No

<!--Notes on breaking changes or related issue tickets if applicable-->

### Development Tested?

- [ ] Yes
- [ ] No

<!--Confirmation of local testing during development-->

## Screenshots
<details>
<summary>Screenshot 1</summary>

<!-- Attach your screenshots here👇-->

</details>

<!--Attach or link to relevant screenshots or visual aids-->

## Test Coverage
<details>
<summary>Screenshot 1</summary>

<!-- Attach your screenshots here👇-->

</details>

<!--Attach Details on test coverage and outcomes-->

## Additional Notes

<!--Any additional notes, considerations or explanations for reviewers-->

tejaskh3 commented:
could you please add a description to this PR?

MayankBansal12 commented on controllers/oooRequests.ts:
```diff
@@ -120,3 +123,37 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+
+    try {
+
+      const response = acknowledgeOOORequest(requestId, requestBody, userId);
```
please use `await` here

surajmaity1 replied:
Hi @MayankBansal12 , it's fixed. 
Thanks

MayankBansal12 commented on test/integration/requests.test.ts:
```diff
@@ -104,9 +104,9 @@ describe("/requests OOO", function () {
         .send(validOooStatusRequests);
-      expect(response).to.have.status(400);
+      expect(response).to.have.status(201);
       expect(response.body).to.have.property("message");
-      expect(response.body.message).to.equal(REQUEST_ALREADY_PENDING);
+      expect(response.body.message).to.equal(REQUEST_CREATED_SUCCESSFULLY);
```
the test title doesn't matches with the response returned here

surajmaity1 replied:
Fixed. For this feature PR, you can find test cases here - https://github.com/Real-Dev-Squad/website-backend/pull/2391

MayankBansal12 commented on middlewares/validators/oooRequests.ts:
```diff
@@ -38,3 +38,35 @@ export const createOooStatusRequestValidator = async (
+    next();
+  } catch (error) {
+    const errorMessages = error.details.map((detail:any) => detail.message);
+    logger.error(`Error while validating request payload : ${errorMessages}`);
```
question: logger is not imported here, how is it being used?

surajmaity1 replied:
It's present in the global object so that we can directly access it without importing.

MayankBansal12 commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,135 @@
+            throw BadRequest(REQUEST_ALREADY_REJECTED);
+        }
+    } catch (error) {
+        logger.error("Error while validating OOO acknowledge request", error);
```
RE: logger not imported

surajmaity1 replied:
https://github.com/Real-Dev-Squad/website-backend/pull/2384#discussion_r2014364477

MayankBansal12 commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,135 @@
+            },
+        };
+    } catch (error) {
+        logger.error("Error while acknowledging OOO request", error);
```
RE: logger not imported

surajmaity1 replied:
https://github.com/Real-Dev-Squad/website-backend/pull/2384#discussion_r2014364477

AnujChhikara commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,135 @@
+                const userFutureStatusData = {
+                    requestId,
+                    status: REQUEST_TYPE.OOO,
+                    state: statusState.UPCOMING,
```
why are we using upcoming here?

AnujChhikara replied:
please reply to this comment

surajmaity1 replied:
Hi @AnujChhikara , it's fixed.
Thank you.

RishiChaubey31 commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,135 @@
+            },
+        };
+    } catch (error) {
+        logger.error("Error while acknowledging OOO request", error);
```
can we please have this error message defined in the constants

surajmaity1 replied:
We're not reusing this error message, so I'm not storing it in the constants.

surajmaity1 commented on controllers/oooRequests.ts:
```diff
@@ -120,3 +123,37 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+
+    const dev = req.query.dev === "true";
+
+    if(!dev) return res.boom.notImplemented("Feature not implemented");
```
This will be removed when the feature flag is removed, so I'm not storing it in the constants.

surajmaity1 commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,135 @@
+                { message: REQUEST_ALREADY_REJECTED }
+            );
+            throw BadRequest(REQUEST_ALREADY_REJECTED);
+        }
```
If we check by PENDING, code will look like this - 
```JS
            if (requestStatus !== REQUEST_STATE.PENDING) {

            let logMessage: string;
            let errorMessage: string;

            if (requestStatus === REQUEST_STATE.APPROVED) {
                logMessage = logType.REQUEST_ALREADY_APPROVED;
                errorMessage = REQUEST_ALREADY_APPROVED;
            }

            if (requestStatus === REQUEST_STATE.REJECTED) {
                logMessage = logType.REQUEST_ALREADY_REJECTED;
                errorMessage = REQUEST_ALREADY_REJECTED;
            }

            await addLog(logMessage,
                { oooRequestId: requestId },
                { message: errorMessage }
            );
            throw BadRequest(errorMessage);
        }
```

Hi @AnujChhikara , please tell me any other way we can fix it or not?

surajmaity1 commented on controllers/oooRequests.ts:
```diff
@@ -120,3 +123,37 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+    const requestId = req.params.id;
+    const isSuperuser = req.userData.roles.super_user === true;
+
+    if (isSuperuser === false) {
```
Fixed.

surajmaity1 commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,135 @@
+
+        if ("error" in requestResult) {
+            throw BadRequest(requestResult.error);
+        }
```
If you check this [line of code](https://github.com/Real-Dev-Squad/website-backend/blob/69eb46578859b98aa8b282885f13b9e6a14e95d3/models/requests.ts#L32), you'll get the idea.

AnujChhikara commented on controllers/oooRequests.ts:
```diff
@@ -120,3 +123,37 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+    const requestBody = req.body;
+    const userId = req.userData.id;
+    const requestId = req.params.id;
+    const isSuperuser = req.userData.roles.super_user === true;
```
```suggestion
    const isSuperuser = req.userData.roles.super_user;
```

surajmaity1 replied:
Fixed.

AnujChhikara commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,135 @@
+        }
+
+        const [acknowledgeLogType, returnMessage] =
+            requestResult.status === REQUEST_STATE.APPROVED
```
Why this check? requestResult does not return the updated request from Firestore; it only returns the body we sent.

surajmaity1 replied:
Correct. At this stage, we verify whether it is approved or rejected to ensure that a specific acknowledgment message is logged and an acknowledgment response is returned.

AnujChhikara replied:
My only concern is that at first glance, requestResult might seem like the updated result from the database, but it's actually just the request body. To avoid confusion, I think it would make more sense to directly use body.status here, as it reflects the actual status being worked with.

surajmaity1 replied:
Yes, that makes sense. We can certainly do that. However, since the updateRequest function returns the request body as-is, I've been using requestResult as both the input and output.

AnujChhikara reviewed (commented):
Hi @surajmaity1,
One of the tests is failing. Could you please take a look when you get a chance?
Thanks!

surajmaity1 commented:
> Hi @surajmaity1,
> One of the tests is failing. Could you please take a look when you get a chance?
> Thanks!


Hi @AnujChhikara ,

The test case for the creation of the OOO request has failed. However, I have already addressed this test case in the PR - "Test PR for Creation of OOO Request."

Thank you.

AnujChhikara commented on middlewares/validators/oooRequests.ts:
```diff
@@ -38,3 +38,46 @@ export const createOooStatusRequestValidator = async (
+      .messages({
+        "any.only": "status must be APPROVED or REJECTED",
+      }),
+    type: joi.string().equal(REQUEST_TYPE.OOO).required().messages({
```
Hello @surajmaity1 I thought we decided not to send request-type in body in the design doc?

surajmaity1 replied:
You can check the discussion as it helps while filtering requests using the type of request.

Please check the [discussion](https://discord.com/channels/673083527624916993/817711078444236821/1355461568682393600)

AnujChhikara commented on middlewares/validators/oooRequests.ts:
```diff
@@ -38,3 +38,46 @@ export const createOooStatusRequestValidator = async (
+        "any.only": "status must be APPROVED or REJECTED",
+      }),
+    type: joi.string().equal(REQUEST_TYPE.OOO).required().messages({
+      "type.any": "type is required",
```
also what happened if I send wrong request type what will be the error response?

surajmaity1 replied:
It'll return - Invalid type request.

AnujChhikara commented on controllers/oooRequests.ts:
```diff
@@ -120,3 +123,37 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+    const requestBody = req.body;
+    const userId = req.userData.id;
+    const requestId = req.params.id;
+    const isSuperuser = req.userData.roles.super_user;
```
super_user field might not always be present. To prevent potential errors, you can use optional chaining here?

AnujChhikara commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,148 @@
+            const requestData = await getRequests({ id: requestId });
+
+            if (requestData) {
+                const { from, until, userId, comment } = requestData as any;
```
Consider avoiding as any for better type safety. You can define a proper type for requestData to ensure strong typing and catch potential errors earlier.

surajmaity1 replied:
Fixed.

AnujChhikara commented:
Hey @surajmaity1, just wanted to check—how do we confirm that the user’s status actually changes after approving the request? I couldn’t find a screenshot or reference for that in the PR. Could you help clarify?

pankajjs commented on controllers/oooRequests.ts:
```diff
@@ -120,3 +123,37 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+    if(!dev) return res.boom.notImplemented("Feature not implemented");
+
+    const requestBody = req.body;
+    const userId = req.userData.id;
```
```suggestion
    const superUserId = req.userData.id;
```

surajmaity1 replied:
Fixed.

Thank you

pankajjs commented on types/oooRequest.d.ts:
```diff
@@ -43,3 +43,19 @@ export type OooRequestResponse = Response & { boom: Boom };
+  dev?: string
+};
+
+export type AcknowledgeOOORequestBody = {
+  comment?: string;
+  status: REQUEST_STATE.APPROVED | REQUEST_STATE.REJECTED;
+}
```
type missing

surajmaity1 replied:
Fixed.

Thank you

pankajjs commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,148 @@
+        await addLog(requestLog.type, requestLog.meta, requestLog.body);
+
+        if (requestResult.status === REQUEST_STATE.APPROVED) {
+            const requestData = await getRequests({ id: requestId });
+
```
this call is redundant

surajmaity1 replied:
Fixed.

Thank you

AnujChhikara commented:
> > Hi @surajmaity1,
> > One of the tests is failing. Could you please take a look when you get a chance?
> > Thanks!
> 
> Hi @AnujChhikara ,
> 
> The test case for the creation of the OOO request has failed. However, I have already addressed this test case in the PR - #2386
> 
> Thank you.

Then how do we plan to merge this into the develop branch? Also, aren't these changes separate from the create route? Since we're working on the patch route, I'm not sure why we're seeing this conflict.

surajmaity1 commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,148 @@
+
+        return {
+            message: returnMessage,
+            data: {
+                id: requestResult.id,
+                ...requestResult,
+            },
```
We are returning the data from the service layer to verify whether the Out-of-Office (OOO) request has been acknowledged by the superuser. This return is primarily for testing and validation purposes. Although the controller does not consume this data directly in the response, it allows us to confirm the internal state during development or debugging. In the actual response sent to the client, we only display a message such as "Request approved/rejected successfully."

surajmaity1 commented:
> > > Hi @surajmaity1,
> > > One of the tests is failing. Could you please take a look when you get a chance?
> > > Thanks!
> > 
> > 
> > Hi @AnujChhikara ,
> > The test case for the creation of the OOO request has failed. However, I have already addressed this test case in the PR - #2386
> > Thank you.
> 
> Then how do we plan to merge this into the develop branch? Also, aren't these changes separate from the create route? Since we're working on the patch route, I'm not sure why we're seeing this conflict.

Hi @AnujChhikara,

The test case is failing because we recently renamed and added a few fields in the OOO request schema. The PR for creating the OOO request already addresses these changes.

So, once that PR is merged, we can proceed to merge this one. That way, we’ll avoid any further failing test cases related to these updates.

Thanks!

surajmaity1 commented on controllers/oooRequests.ts:
```diff
@@ -120,3 +123,37 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+
+    if (!isSuperuser) {
+      return res.boom.unauthorized(UNAUTHORIZED_TO_ACKNOWLEDGE_OOO_REQUEST);
+    }
```
Fixed.

Thank you

pankajjs commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,146 @@
+        await addLog(requestLog.type, requestLog.meta, requestLog.body);
+
+        if (requestResult.status === REQUEST_STATE.APPROVED) {
+            if (requestData) {
```
do we need this if check?

MayankBansal12 commented on controllers/oooRequests.ts:
```diff
@@ -30,7 +37,7 @@ export const createOooRequestController = async (req: OooRequestCreateRequest, r
     const latestOooRequest:OooStatusRequest = await getRequestByKeyValues({ requestedBy: userId, type: REQUEST_TYPE.OOO , state: REQUEST_STATE.PENDING });
 
-    if (latestOooRequest && latestOooRequest.state === REQUEST_STATE.PENDING) {
+    if (latestOooRequest && latestOooRequest.status === REQUEST_STATE.PENDING) {
```
just wanted a clarification: the field name change for `state` to `status` is still happening?

surajmaity1 replied:
yes we're renaming state field to status

MayankBansal12 replied:
if this change isn't happening in your PRs then i would suggest sticking to `state` field here as it will help you in testing during staging and prod and let this field name change happen in this [issue](https://github.com/Real-Dev-Squad/website-backend/issues/2399)

surajmaity1 replied:
As per our design document discussion, we had concluded that the field rename from state to status would begin with this implementation. Additionally, Aditya has been assigned the corresponding issue and will handle updating the remaining references accordingly.

MayankBansal12 replied:
if you feel this won't create issues in testing, then it's fine by me. you can resolve the conversation

MayankBansal12 commented on middlewares/validators/oooRequests.ts:
```diff
@@ -38,3 +38,46 @@ export const createOooStatusRequestValidator = async (
+    await schema.validateAsync(req.body, { abortEarly: false });
+    next();
+  } catch (error) {
+    const errorMessages = error.details.map((detail:any) => detail.message);
```
can we avoid using any here? `detail:any`

surajmaity1 replied:
It's fixed.
Thank you

MayankBansal12 commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,158 @@
+
+    try {
+
+        if (requestType !== REQUEST_TYPE.OOO) {
+            await addLog(logType.INVALID_REQUEST_TYPE,
+                { requestId, type: requestType },
+                { message: INVALID_REQUEST_TYPE }
+            );
+            return {
+                error: INVALID_REQUEST_TYPE
+            };
+        }
```
question: if we are using validator to validate request type for OOO, why do we need this check here?

surajmaity1 replied:
In the request body, we are passing a type field. As observed on [line 88 of service/oooRequest.ts](https://github.com/Real-Dev-Squad/website-backend/pull/2384/files#diff-2f21c5728813e485d091602c3581d748d1e0ac12ebd56c674caa7f274268f066R88), we are retrieving the request from the database using the request ID. It is important to ensure that the type provided in the request body matches the type of the request fetched from the database.

MayankBansal12 commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,158 @@
+            };
+        }
+
+        if (requestStatus === REQUEST_STATE.APPROVED) {
+            await addLog(logType.REQUEST_ALREADY_APPROVED,
+                { oooRequestId: requestId },
+                { message: REQUEST_ALREADY_APPROVED }
+            );
+            return {
+                error: REQUEST_ALREADY_APPROVED
+            };
+        }
+
+        if (requestStatus === REQUEST_STATE.REJECTED) {
+            await addLog(logType.REQUEST_ALREADY_REJECTED,
+                { oooRequestId: requestId },
+                { message: REQUEST_ALREADY_REJECTED }
+            );
+            return {
+                error: REQUEST_ALREADY_REJECTED
+            };
+        }
```
please refactor this, we are doing the same thing here, if status is already changed `APPROVED` or `REJECTED` throw the error

MayankBansal12 commented on services/oooRequest.ts:
```diff
@@ -0,0 +1,158 @@
+ * @param {string} requestStatus - The current status of the request.
+ * @throws {Error} Throws an error if an issue occurs during validation.
+ */
+export const validateOOOAcknowledgeRequest = async (
```
if the only use of this func is to ensure that the OOO request is PENDING then we should consider using a simple naming, this one was confusing and took me a while to understand

surajmaity1 replied:
In the request body, we are passing a type field. As observed on [line 88 of service/oooRequest.ts](https://github.com/Real-Dev-Squad/website-backend/pull/2384/files#diff-2f21c5728813e485d091602c3581d748d1e0ac12ebd56c674caa7f274268f066R88), we are retrieving the request from the database using the request ID. It is important to ensure that the type provided in the request body matches the type of the request fetched from the database.

prakashchoudhary07 commented on controllers/oooRequests.ts:
```diff
@@ -120,3 +127,56 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+    const requestId = req.params.id;
+    const isSuperuser = req.userData.roles?.super_user;
+
+    if (!isSuperuser) {
+      return res.boom.forbidden(UNAUTHORIZED_TO_ACKNOWLEDGE_OOO_REQUEST);
+    }
+
```
Why is this check here?
Will this not happen in middleware? If its a route only super user can access?

surajmaity1 replied:
This route is intended to be accessed by both superusers and Discord users. However, only superusers are authorized to approve or reject Out of Office (OOO) requests. While this authorization check could be implemented in middleware, doing so would apply the restriction to all requests to this route.

In summary, both superusers and regular users can access the route, but only superusers are permitted to perform certain actions like acknowledging OOO requests.

MayankBansal12 replied:
@surajmaity1 this seems confusing, why should normal users access the acknowledge request route, and for what purpose?

can you please explain if I'm missing something here?

surajmaity1 replied:
Hi @MayankBansal12 , I updated reply. Can you please check.

AnujChhikara commented on services/oooRequest.ts:
```diff
@@ -93,3 +106,141 @@ export const createOooRequest = async (
+        }
+
+        if (requestStatus === REQUEST_STATE.REJECTED) {
+            await addLog(logType.REQUEST_ALREADY_REJECTED,
```
variable name is wrong why we still using REQUEST_STATE

surajmaity1 replied:
All kind of request ( such as OOO, ONBOARDING ) are using this REQUEST_STATE.
We'll change this variable name while doing code cleanup in the future.

AnujChhikara commented on services/oooRequest.ts:
```diff
@@ -93,3 +106,141 @@ export const createOooRequest = async (
+            };
+        }
+
+        const requestResult = await updateRequest(requestId, body, superUserId, REQUEST_TYPE.OOO);
```
Please consider updating the `updateRequest` function — it's still checking for the state field. Since the field has been renamed to status, it would be good to update it here for consistency and correctness.

surajmaity1 replied:
We already checked in this `validateOooAcknowledgeRequest` function, we're checking it's APPROVED or REJECTED inside this function. 

We'll fix this `updateRequest` function in #2399 .

Thank you

surajmaity1 commented on middlewares/validators/oooRequests.ts:
```diff
@@ -38,3 +38,46 @@ export const createOooStatusRequestValidator = async (
+  .object()
+  .strict()
+  .keys({
+    comment: joi.string().optional()
+      .messages({
+        "string.empty": "comment cannot be empty",
+      }),
```
It means the `comment `field is optional, but if it is provided, it must not be an empty string.

Why this makes sense:
- `.optional() `→ The field is not required; it can be missing from the payload entirely.

- `.messages({ "string.empty": "comment cannot be empty" })` → If the field is present, and is an empty string (""), that’s invalid, and this message will be shown.

surajmaity1 commented on controllers/oooRequests.ts:
```diff
@@ -148,3 +154,56 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+    const isSuperuser = req.userData.roles?.super_user;
+
+    if (!isSuperuser) {
+      return res.boom.forbidden(UNAUTHORIZED_TO_ACKNOWLEDGE_OOO_REQUEST);
```
Only superusers are authorized to approve or reject OOO requests. If a non-superuser attempts to perform this action, the system will display the following message: `Only superusers are allowed to acknowledge OOO requests.`

surajmaity1 commented on services/oooRequest.ts:
```diff
@@ -5,12 +5,25 @@ import {
+import { updateRequest } from "../models/requests";
+import { AcknowledgeOooRequestBody } from "../types/oooRequest";
+import { addFutureStatus } from "../models/userStatus";
+const requestModel = firestore.collection("requests");
```
We are accessing the collection directly within the service because there is currently no dedicated function in the model layer to retrieve a single request by its ID. As a result, we are using `requestModel.doc(requestId).get()` within the service to fetch the request data directly.

surajmaity1 commented on services/oooRequest.ts:
```diff
@@ -93,3 +106,141 @@ export const createOooRequest = async (
+
+        const requestData = request.data();
+
+        const validationResponse = await validateOooAcknowledgeRequest(requestId, requestData.type, requestData.status);
+
+        if (validationResponse) {
+            return {
+                error: validationResponse.error
+            };
+        }
```
If validationResponse throws an error, could you please suggest the appropriate process to handle it?

Would it be acceptable to approach it like this:
```ts
export const validateOooAcknowledgeRequest = async (
  requestId: string,
  requestType: string,
  requestStatus: string,
) => {
  if (requestType !== REQUEST_TYPE.OOO) {
    await addLog(logType.INVALID_REQUEST_TYPE,
      { requestId, type: requestType },
      { message: INVALID_REQUEST_TYPE }
    );
    throw new Error(INVALID_REQUEST_TYPE);
  }

  if (requestStatus === REQUEST_STATE.APPROVED) {
    await addLog(logType.REQUEST_ALREADY_APPROVED,
      { oooRequestId: requestId },
      { message: REQUEST_ALREADY_APPROVED }
    );
    throw new Error(REQUEST_ALREADY_APPROVED);
  }

  if (requestStatus === REQUEST_STATE.REJECTED) {
    await addLog(logType.REQUEST_ALREADY_REJECTED,
      { oooRequestId: requestId },
      { message: REQUEST_ALREADY_REJECTED }
    );
    throw new Error(REQUEST_ALREADY_REJECTED);
  }
}


```

surajmaity1 commented on test/fixtures/oooRequest/oooRequest.ts:
```diff
@@ -168,7 +168,7 @@ export const createOooRequests3 = {
 };
 
-export const acknowledgeOooRequest = {
+export const acknowledgeOooRequestBody = {
```
I'll fix it.
Thank you

surajmaity1 commented on middlewares/validators/oooRequests.ts:
```diff
@@ -38,3 +38,46 @@ export const createOooStatusRequestValidator = async (
+    type: joi.string().equal(REQUEST_TYPE.OOO).required().messages({
+      "type.any": "type is required",
+    })
+  });
```
Hi Yash,
It's fixed now.
Thank you!

surajmaity1 commented on services/oooRequest.ts:
```diff
@@ -93,3 +106,141 @@ export const createOooRequest = async (
+            await addLog(logType.INVALID_REQUEST_TYPE,
+                { requestId, type: requestType },
+                { message: INVALID_REQUEST_TYPE }
+            );
```
Hi Yash,
It's fixed now.
Thank you!

surajmaity1 commented on services/oooRequest.ts:
```diff
@@ -93,3 +106,141 @@ export const createOooRequest = async (
+        const request = await requestModel.doc(requestId).get();
+
+        if (!request.exists) {
+            await addLog(logType.REQUEST_DOES_NOT_EXIST,
```
Hi Yash,
It's fixed now.
Thank you!

pankajjs commented on test/unit/services/oooRequest.test.ts:
```diff
@@ -183,64 +183,64 @@ describe("Test OOO Request Service", function() {
-            //     expect(createSpy.calledOnce).to.be.true;
-            // }
+            sinon.stub(logService, "addLog").throws(new Error(errorMessage));
+            const createSpy = sinon.spy(require("../../../services/oooRequest"), "acknowledgeOooRequest");
```
- Why are we spying here?
 - Dependencies are not mocked.

pankajjs commented on test/unit/services/oooRequest.test.ts:
```diff
@@ -183,64 +183,64 @@ describe("Test OOO Request Service", function() {
-            //         updatedAt: response.data.updatedAt
-            //     }
-            // });
+            const response = await acknowledgeOooRequest(
```
- Dependencies are not mocked.

vikasosmium replied:
Not required here for Internal dependencies

pankajjs commented on test/unit/services/oooRequest.test.ts:
```diff
@@ -183,64 +183,64 @@ describe("Test OOO Request Service", function() {
-            //         updatedAt: response.data.updatedAt
-            //     }
-            // });
+            const response = await acknowledgeOooRequest(
```
- Dependencies are not mocked.

vikasosmium replied:
Not required here for Internal dependencies

pankajjs commented on test/unit/services/oooRequest.test.ts:
```diff
@@ -183,64 +183,64 @@ describe("Test OOO Request Service", function() {
         });
 
         it("should return REQUEST_DOES_NOT_EXIST if invalid request id is passed", async function () {
-            // const invalidOOORequestId = "11111111111111111111";
```
- Dependencies are not mocked.

vikasosmium replied:
not required here I think

pankajjs commented on services/oooRequest.ts:
```diff
@@ -93,3 +104,125 @@ export const createOooRequest = async (
+    superUserId: string,
+) => {
+    try {
+        const requestData = await getRequest(requestId);
+
+        if (requestData.error) {
+            logger.error("Error while acknowledging OOO request", { requestId, reason: REQUEST_DOES_NOT_EXIST });
+            return {
+                error: REQUEST_DOES_NOT_EXIST
+            };
+        }
+
```
- Why are not throwing error from getRequest?
- Can we rename the function to getRequestById?

surajmaity1 replied:
I renamed the function from getRequest to getRequestById.

pankajjs commented on services/oooRequest.ts:
```diff
@@ -93,3 +104,125 @@ export const createOooRequest = async (
+            };
+        }
+
+        const validationError = await validateOooAcknowledgeRequest(requestId, requestData.type, requestData.status);
+
+        if (validationError) {
+            logger.error(`Error while validating OOO acknowledge request`, { requestId, reason: validationError.error });
+            return {
+                error: validationError.error
+            };
+        }
```
- Why are we not throwing error from validateOooAcknowledgeRequest?

pankajjs commented on controllers/oooRequests.ts:
```diff
@@ -148,3 +154,56 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+    const requestBody = req.body;
+    const superUserId = req.userData.id;
+    const requestId = req.params.id;
+    const isSuperuser = req.userData.roles?.super_user;
```
These lines can be placed after super user check?

pankajjs commented on controllers/oooRequests.ts:
```diff
@@ -148,3 +154,56 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+
+      const response = await acknowledgeOooRequest(requestId, requestBody, superUserId);
+
+      if (response.error === REQUEST_DOES_NOT_EXIST) {
+        return res.boom.notFound(REQUEST_DOES_NOT_EXIST);
+      }
+      if (response.error === INVALID_REQUEST_TYPE) {
+        return res.boom.badRequest(INVALID_REQUEST_TYPE);
+      }
+      if (response.error === REQUEST_ALREADY_APPROVED) {
+        return res.boom.conflict(REQUEST_ALREADY_APPROVED);
+      }
+      if (response.error === REQUEST_ALREADY_REJECTED) {
+        return res.boom.conflict(REQUEST_ALREADY_REJECTED);
+      }
+
```
This should be handled in service layer.

pankajjs commented on controllers/oooRequests.ts:
```diff
@@ -148,3 +154,56 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+    }
+    catch(error){
+      logger.error(ERROR_WHILE_ACKNOWLEDGING_REQUEST, error);
+      return res.boom.badImplementation(ERROR_WHILE_ACKNOWLEDGING_REQUEST);
```
Why are we not calling error handler from here?

pankajjs commented on models/requests.ts:
```diff
@@ -69,6 +69,25 @@ export const updateRequest = async (id: string, body: any, lastModifiedBy: strin
+  try {
+    const requestDoc = await requestModel.doc(id).get();
+
+    if (!requestDoc.exists) {
+      return {
+        error: REQUEST_DOES_NOT_EXIST,
+      };
+    }
+
```
This should throw error.

pankajjs commented on models/requests.ts:
```diff
@@ -69,6 +69,25 @@ export const updateRequest = async (id: string, body: any, lastModifiedBy: strin
+      };
+    }
+
+    const request = requestDoc.data();
+
+    return request;
```
```suggestion
    return requestDoc.data();
```

surajmaity1 replied:
fixed.

pankajjs commented on services/oooRequest.ts:
```diff
@@ -93,3 +104,125 @@ export const createOooRequest = async (
+
+    try {
+
+        if (requestType !== REQUEST_TYPE.OOO) {
+            return {
+                error: INVALID_REQUEST_TYPE
+            };
+        }
+
+        if (requestStatus === REQUEST_STATE.APPROVED) {
+            return {
+                error: REQUEST_ALREADY_APPROVED
+            };
+        }
+
+        if (requestStatus === REQUEST_STATE.REJECTED) {
+            return {
+                error: REQUEST_ALREADY_REJECTED
+            };
+        }
```
Please throw respective errors.

pankajjs commented on services/oooRequest.ts:
```diff
@@ -93,3 +104,125 @@ export const createOooRequest = async (
+                requestId: requestId,
+                action: LOG_ACTION.UPDATE,
+                userId: superUserId,
+                createdAt: Date.now(),
```
the addLog function have default timestamp field.
- createdAt field is redundant here.

pankajjs reviewed (changes requested):
Proof missing for case when after approving request, user status becomes OOO.

surajmaity1 commented:
> Proof missing for case when after approving request, user status becomes OOO.

After the OOO request is approved, the user's `futureStatus `is updated accordingly. A scheduled cron job then processes this status change and updates the user's status to "OOO" at the appropriate time.

pankajjs reviewed (approved):
LGTM

vinit717 commented on models/requests.ts:
```diff
@@ -69,6 +70,21 @@ export const updateRequest = async (id: string, body: any, lastModifiedBy: strin
   }
 };
 
+export const getRequestById = async (id: string) => {
```
Where is the test for this part?

surajmaity1 replied:
Here are [test cases](https://github.com/Real-Dev-Squad/website-backend/pull/2384/files#diff-445f1d4f3e2faa636a8e1d79f40546750f52332535f90f3244c9381728ff5be7) for this function

vinit717 commented on middlewares/validators/oooRequests.ts:
The middleware test is missing? Also, why do we need to separate a middleware file for ooo requests?

surajmaity1 replied:
Middleware tests are written in this test PR
- https://github.com/Real-Dev-Squad/website-backend/pull/2391

We're creating each middleware for all types of requests ( such as ONBOARDING, OOO etc. )

vinit717 commented on controllers/oooRequests.ts:
here also why we need seprate file for ooo requests?

surajmaity1 replied:
We are implementing dedicated controllers for each request type (e.g., ONBOARDING, OOO, etc.) so that incoming requests are routed to the appropriate handler based on their type.

vinit717 commented on controllers/oooRequests.ts:
```diff
@@ -148,3 +151,45 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+ * @param {OooRequestResponse} res - The response object.
+ * @returns {Promise<OooRequestResponse>} Resolves with success or failure.
+ */
+export const acknowledgeOooRequestController = async (
```
fix the function name

surajmaity1 replied:
Hi @vinit717 ,
What is the name you're suggesting?

vinit717 replied:
This is the controller file that why there is controller name in this function

vinit717 commented on models/requests.ts:
```diff
@@ -69,6 +70,21 @@ export const updateRequest = async (id: string, body: any, lastModifiedBy: strin
   }
 };
 
+export const getRequestById = async (id: string) => {
```
Where is the test for this function?

surajmaity1 replied:
Here are[ test cases](https://github.com/Real-Dev-Squad/website-backend/pull/2384/files#diff-445f1d4f3e2faa636a8e1d79f40546750f52332535f90f3244c9381728ff5be7) for this function

iamitprakash commented on controllers/oooRequests.ts:
```diff
@@ -148,3 +152,45 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+
+    if(!dev) return res.boom.notImplemented("Feature not implemented");
+
+    const isSuperuser = req.userData.roles?.super_user;
```
please add optional chaining to req.userData?.roles to prevent potential null/undefined errors if userData isn't properly initialized.

surajmaity1 replied:
Hi Amit sir,
it's fixed.

iamitprakash commented on controllers/oooRequests.ts:
```diff
@@ -148,3 +152,45 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+
+    const requestBody = req.body;
+    const superUserId = req.userData.id;
+    const requestId = req.params.id;
```
please add validation for requestId

surajmaity1 replied:
Hi Amit sir,
it's fixed.

iamitprakash commented on controllers/oooRequests.ts:
```diff
@@ -148,3 +152,45 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
+ * @param {OooRequestResponse} res - The response object.
+ * @returns {Promise<OooRequestResponse>} Resolves with success or failure.
+ */
+export const acknowledgeOooRequest = async (
```
NIT: please move the entire function body inside the try-catch block to capture all potential errors, including those that might occur during parameter extraction

surajmaity1 replied:
Hi Amit sir,
it's fixed.

surajmaity1 commented:
> @surajmaity1 please use below
> 
> ```
> export const acknowledgeOooRequest = async (
>   req: AcknowledgeOooRequest,
>   res: OooRequestResponse,
>   next: NextFunction
> ): Promise<OooRequestResponse> => {
>   try {
>     const dev = req.query.dev === "true";
>     if (!dev) return res.boom.notImplemented("Feature not implemented");
>     
>     const isSuperuser = req.userData?.roles?.super_user;
>     if (!isSuperuser) {
>       return res.boom.forbidden(UNAUTHORIZED_TO_UPDATE_REQUEST);
>     }
>     
>     const requestBody = req.body;
>     const superUserId = req.userData.id;
>     const requestId = req.params.id;
>     
>     if (!requestId) {
>       return res.boom.badRequest("Request ID is required");
>     }
>     
>     const response = await oooRequestService.acknowledgeOooRequest(requestId, requestBody, superUserId);
>     return res.status(200).json({
>       message: response.message,
>     });
>   }
>   catch (error) {
>     logger.error(ERROR_WHILE_ACKNOWLEDGING_REQUEST, error);
>     next(error);
>     return res;
>   }
> };
> ```

Hi Amit sir,
it's added.
